substrate/frame/election-provider-multi-block/src/lib.rs

```diff
@@ -708,51 +713,71 @@ pub mod pallet {
 
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
-		fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
-			let current_phase = CurrentPhase::<T>::get();
-			let weight1 = match current_phase {
-				Phase::Snapshot(x) if x == T::Pages::get() => {
-					// create the target snapshot
-					Self::create_targets_snapshot();
-					T::WeightInfo::on_initialize_into_snapshot_msp()
-				},
-				Phase::Snapshot(x) => {
-					// create voter snapshot
-					Self::create_voters_snapshot_paged(x);
-					T::WeightInfo::on_initialize_into_snapshot_rest()
-				},
-				_ => T::WeightInfo::on_initialize_nothing(),
-			};
-
-			// Only transition if not in Export phase
-			if !matches!(current_phase, Phase::Export(_)) {
-				let next_phase = current_phase.next();
+		fn on_poll(_now: BlockNumberFor<T>, weight_meter: &mut WeightMeter) {
+			// first check we can at least read one storage.
+			if !weight_meter.can_consume(T::DbWeight::get().reads(1)) {
+				Self::deposit_event(Event::UnexpectedPhaseTransitionHalt {
+					required: T::DbWeight::get().reads(1),
+					had: weight_meter.remaining(),
+				});
+				return;
+			}
 
-				let weight2 = match next_phase {
-					Phase::Signed(_) => T::WeightInfo::on_initialize_into_signed(),
-					Phase::SignedValidation(_) =>
-						T::WeightInfo::on_initialize_into_signed_validation(),
-					Phase::Unsigned(_) => T::WeightInfo::on_initialize_into_unsigned(),
-					_ => T::WeightInfo::on_initialize_nothing(),
-				};
+			// if so, consume and prepare the next phase.
+			let current_phase = Self::current_phase();
+			let next_phase = current_phase.next();
+			weight_meter.consume(T::DbWeight::get().reads(1));
+
+			let (self_weight, self_exec) = Self::per_block_exec(current_phase);
```

Comment:
Whats the advantage of returning the closure instead of executing the logic directly?

Reply:
I am not sure if it helps with readability.